src/main/kotlin/com/mineinabyss/deeperworld/synchronization/ExploitPreventionListener.kt

```diff
@@ -1,24 +1,45 @@
 package com.mineinabyss.deeperworld.synchronization
 
-import com.mineinabyss.deeperworld.world.section.correspondingSection
+import com.mineinabyss.deeperworld.world.section.correspondingLocation
+import com.mineinabyss.deeperworld.world.section.inSectionOverlap
+import io.papermc.paper.event.block.BlockBreakBlockEvent
+import org.bukkit.block.Block
 import org.bukkit.event.EventHandler
 import org.bukkit.event.Listener
 import org.bukkit.event.block.BlockPistonExtendEvent
 import org.bukkit.event.block.BlockPistonRetractEvent
+import org.bukkit.inventory.ItemStack
 
 object ExploitPreventionListener : Listener {
-    /** Disables pistons extending if they are in the overlap of two sections */
+
+    @EventHandler
+    fun BlockPistonExtendEvent.onPistonExtendEvent() {
+        val corrBlock = block.location.correspondingLocation?.block ?: return
+        val corrBlocks = mutableListOf<Block>()
+
+        if (blocks.all { b -> !b.location.inSectionOverlap }) return
+        if (!corrBlock.location.isChunkLoaded) corrBlock.chunk.load()
+
+        blocks.forEach { corrBlocks.add(it.location.correspondingLocation?.block ?: return@forEach) }
+        BlockPistonRetractEvent(corrBlock, corrBlocks, direction).callEvent()
+    }
+
     @EventHandler
-    fun onPistonExtendEvent(event: BlockPistonExtendEvent) {
-        //TODO handle pistons properly instead of just cancelling the event
-        if (event.blocks.any { it.location.correspondingSection != null })
-            event.isCancelled = true
+    fun BlockPistonRetractEvent.onPistonRetractEvent() {
+        val corrBlock = block.location.correspondingLocation?.block ?: return
+        val corrBlocks = mutableListOf<Block>()
+
+        if (blocks.all { b -> !b.location.inSectionOverlap }) return
+        if (!corrBlock.location.isChunkLoaded) corrBlock.chunk.load()
+
+        blocks.forEach { corrBlocks.add(it.location.correspondingLocation?.block ?: return@forEach) }
+        BlockPistonRetractEvent(corrBlock, corrBlocks, direction).callEvent()
     }
 
-    /** Disables pistons retracting if they are in the overlap of two sections */
     @EventHandler
-    fun onPistonRetractEvent(event: BlockPistonRetractEvent) {
-        if (event.blocks.any { it.location.correspondingSection != null })
-            event.isCancelled = true
+    fun BlockBreakBlockEvent.onPistonBreakBlock() {
+        val corrBlock = block.location.correspondingLocation?.block ?: return
+        val corrSource = source.location.correspondingLocation?.block ?: return
+        BlockBreakBlockEvent(corrBlock, corrSource, emptyList<ItemStack>()) // Empty drops to avoid dupes
```

Comment:
You are not calling this event like above.

Also i do not think events work like this, you would manually need to trigger a block break if i remember correctly. The event is only to trigger listeners.

Reply:
Yeah i think it just worked and didnt look over, would manually removing the block make more sense? just set it to air
should make more sense aswell?